cub/cub/agent/agent_segmented_scan.cuh

```diff
@@ -0,0 +1,296 @@
+struct AgentSegmentedScanPolicy : ScalingType
+{
+  static constexpr BlockLoadAlgorithm LOAD_ALGORITHM   = _LOAD_ALGORITHM;
+  static constexpr CacheLoadModifier LOAD_MODIFIER     = _LOAD_MODIFIER;
+  static constexpr BlockStoreAlgorithm STORE_ALGORITHM = _STORE_ALGORITHM;
+  static constexpr BlockScanAlgorithm SCAN_ALGORITHM   = _SCAN_ALGORITHM;
+};
+
+/******************************************************************************
+ * Thread block abstractions
+ ******************************************************************************/
+
+namespace detail::segmented_scan
+{
+
+/**
+ * @brief AgentSegmentedScan implements a stateful abstraction of CUDA thread blocks for
+ *        participating in device-wide segmented prefix scan.
+ * @tparam AgentSegmentedScanPolicyT
+ *   Parameterized AgentSegmentedScanPolicyT tuning policy type
+ *
+ * @tparam InputIteratorT
+ *   Random-access input iterator type
+ *
+ * @tparam OutputIteratorT
+ *   Random-access output iterator type
+ *
+ * @tparam OffsetT
+ *   Signed integer type for global offsets
+ *
+ * @tparam ScanOpT
+ *   Scan functor type
+ *
+ * @tparam InitValueT
+ *   The init_value element for ScanOpT type (cub::NullType for inclusive scan)
+ *
+ * @tparam AccumT
+ *   The type of intermediate accumulator (according to P2322R6)
+ */
+template <typename AgentSegmentedScanPolicyT,
+          typename InputIteratorT,
+          typename OutputIteratorT,
+          typename OffsetT,
+          typename ScanOpT,
+          typename InitValueT,
+          typename AccumT,
+          bool ForceInclusive = false>
+struct AgentSegmentedScan
+{
+  //---------------------------------------------------------------------
+  // Types and constants
+  //---------------------------------------------------------------------
+
+  // The input value type
+  using InputT = cub::detail::it_value_t<InputIteratorT>;
+
+  // Input iterator wrapper type (for applying cache modifier)
+  // Wrap the native input pointer with CacheModifiedInputIterator
+  // or directly use the supplied input iterator type
+  using WrappedInputIteratorT =
+    ::cuda::std::_If<::cuda::std::is_pointer_v<InputIteratorT>,
+                     CacheModifiedInputIterator<AgentSegmentedScanPolicyT::LOAD_MODIFIER, InputT, OffsetT>,
+                     InputIteratorT>;
+
+  // Constants
+  enum
+  {
+    // Use cub::NullType means no initial value is provided
+    HAS_INIT = !::cuda::std::is_same_v<InitValueT, NullType>,
+    // We are relying on either initial value not being `NullType`
+    // or the ForceInclusive tag to be true for inclusive scan
+    // to get picked up.
+    IS_INCLUSIVE     = ForceInclusive || !HAS_INIT,
+    BLOCK_THREADS    = AgentSegmentedScanPolicyT::BLOCK_THREADS,
+    ITEMS_PER_THREAD = AgentSegmentedScanPolicyT::ITEMS_PER_THREAD,
+    TILE_ITEMS       = BLOCK_THREADS * ITEMS_PER_THREAD,
+  };
+
+  // Parametrized BlockLoad type
+  using BlockLoadT =
```

Comment:
Strong suggestion: please avoid comments that don't add any new information to the code. This comment does not add any value IMO.